### PR TITLE
[WIP] Example Code Blocks and Styles

### PR DESCRIPTION
**Pull request by JHibbard**
This PR is a work in progress to show potential components for code blocks and documentation callouts.

Screen Shot 2022-05-18 at 1 30 54 AM

### CODE REVIEW

**Comment by dennyglee** [comment]
FYI - please compare this version with the original docs.delta.io version at: https://docs.delta.io/latest/delta-batch.html

File: src/pages/learn/batch.mdx [hunk 1]
```diff
@@ -0,0 +1,463 @@
+---
+title: Delta Operations
+width: full
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:3]: Is the plan to have the side submenu for the different sections?  Thanks!

File: src/pages/learn/batch.mdx [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+
+For many <DeltaLake /> operations on tables, you enable integration with
+<ApacheSpark /> DataSourceV2 and Catalog APIs (since 3.0) by setting configurations
+when you create a new <SparkSession />.
+
+## Create a Table
+
+You can create tables in the following ways:
+
+- SQL DDL commands: You can use standard SQL DDL commands supported in Apache Spark (for example, CREATE TABLE and REPLACE TABLE) to create Delta tables.
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:30]: Please bold "SQL DDL commands"

File: src/pages/learn/batch.mdx [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+      ssn STRING,
+      salary INT
+    ) USING DELTA
+    ```
+
+    </div>
+      }
+    />
+
+- DataFrameWriter API: If you want to simultaneously create a table and insert data into it from Spark DataFrames or Datasets, you can use the Spark DataFrameWriter (Scala or Java and Python).
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:88]: Please bold "DataFrameWriter API"

File: src/pages/learn/batch.mdx [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    df.write.format("delta").mode("overwrite").save("/tmp/delta/people10m")
+    ```
+
+    </div>
+      }
+    />
+
+    You can also create Delta tables using the Spark DataFrameWriterV2 API.
+
+- DeltaTableBuilder API: You can also use the DeltaTableBuilder API in Delta Lake to create tables. Compared to the DataFrameWriter APIs, this API makes it easier to specify additional information like column comments, table properties, and generated columns.
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:121]: Please bold "DeltaTableBuilder API"

File: src/pages/learn/batch.mdx [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    .addColumn("salary", "INT")
+    .partitionedBy("gender")
+    .execute()
+  ```
+
+  </div>
+  }
+/>
+
+To determine whether a table contains a specific partition, use the statement SELECT COUNT(*) &gt; 0 FROM &lt;table-name&gt; WHERE &lt;partition-column&gt; = &lt;value&gt;. If the partition exists, true is returned. For example:
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:287]: Please put `<code></code>` around the SELECT statement, i.e. `SELECT COUNT(*) > 0 FROM <table-name> WHERE <partition-column> = <value>`

File: src/pages/learn/batch.mdx [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+  display(spark.sql("SELECT COUNT(*) > 0 AS `Partition exists` FROM default.people10m WHERE gender = 'M'"))
+  ```
+
+  </div>
+  }
+/>
+
+## Control Data Location
+
+For tables defined in the metastore, you can optionally specify the LOCATION as a path. Tables created with a specified LOCATION are considered unmanaged by the metastore. Unlike a managed table, where no path is specified, an unmanaged table’s files are not deleted when you DROP the table.
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:321]: Please put `<code></code>` around `LOCATION`, `DROP`

File: src/pages/learn/batch.mdx [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+
+When you run CREATE TABLE with a LOCATION that already contains data stored using Delta Lake, Delta Lake does the following:
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:323]: Please put `<code></code>` around `CREATE TABLE`, `LOCATION` and italicize *already*.

File: src/pages/learn/batch.mdx [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+
+- If you specify only the table name and location, for example:
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:325]: Please italicize *only the table name and location*

File: src/pages/learn/batch.mdx [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+    LOCATION '/tmp/delta/people10m'
+    ```
+
+    </div>
+    }
+  />
+
+  the table in the metastore automatically inherits the schema, partitioning, and table properties of the existing data. This functionality can be used to “import” data into the metastore.
+
+- If you specify any configuration (schema, partitioning, or table properties), Delta Lake verifies that the specification exactly matches the configuration of the existing data.
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:343]: Please italicize *any configuration*

File: src/pages/learn/batch.mdx [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+</Info>
+
+## Use Generated Columns
+
+<Info title="Note" level="info">
+  This feature is new and is in Preview.
+</Info>
+
+Delta Lake supports generated columns which are a special type of columns whose values are automatically generated based on a user-specified function over other columns in the Delta table. When you write to a table with generated columns and you do not explicitly provide values for them, Delta Lake automatically computes the values. For example, you can automatically generate a date column (for partitioning the table by date) from the timestamp column; any writes into the table need only specify the data for the timestamp column. However, if you explicitly provide values for them, the values must satisfy the constraint (&lt;value&gt; &lt;=&gt; &lt;generation expression&gt;) IS TRUE or the write will fail with an error.
```
Comment (dennyglee) [inline src/pages/learn/batch.mdx:360]: Please specify `<code></code>` around `constraint (<value> <=> <generation expression>) IS TRUE`

**Review by dennyglee — CHANGES_REQUESTED** [review]
Could you review the `<code></code>` and formatting of the text?  Just added some quick call outs - thanks!

File: src/components/CodeBlocks.jsx [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
+import 'react-tabs/style/react-tabs.css';
+import * as React from "react";
+
+const CBCreateTable = ({sql, python, scala}) => {
+    return (
+        <Tabs>
+            <TabList>
+                {!sql || <Tab>SQL</Tab>}
```
Comment (scottsand-db) [inline src/components/CodeBlocks.jsx:9]: nit: is it just me, or isn't `sql && <Tab>SQL</Tab>` a little logically easier to reason about?

File: src/components/CodeBlocks.jsx [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
+import 'react-tabs/style/react-tabs.css';
+import * as React from "react";
+
+const CBCreateTable = ({sql, python, scala}) => {
```
Comment (scottsand-db) [inline src/components/CodeBlocks.jsx:5]: what does `CB` mean?

Comment (scottsand-db) [reply src/components/CodeBlocks.jsx:5]: i guess CodeBlocks? Is prefixing this jsx class with `CB` a convention? does it make it more clear or readable?

File: src/pages/learn/batch.mdx [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+    ssn STRING,
+    salary INT
+  )
+  USING DELTA
+  PARTITIONED BY (gender)
+  ```
+
+  </div>
+  }
+  python={
```
Comment (scottsand-db) [inline src/pages/learn/batch.mdx:236]: shouldn't there be a comma after `}` ? also, again, here indentiation would help to clearly separate the SQL code from the PYTHON code.

File: src/components/CodeBlocks.jsx (hunk 12 above)
Comment (scottsand-db) [inline src/components/CodeBlocks.jsx:5]: nit: class docs, just a 1 liner about what this class does and the expected inputs.

File: src/pages/learn/batch.mdx [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+- Two partition columns defined by YEAR(col), MONTH(col) and the type of col is TIMESTAMP.
+- Three partition columns defined by YEAR(col), MONTH(col), DAY(col) and the type of col is TIMESTAMP.
+- Four partition columns defined by YEAR(col), MONTH(col), DAY(col), HOUR(col) and the type of col is TIMESTAMP.
+- SUBSTRING(col, pos, len) and the type of col is STRING
+- DATE_FORMAT(col, format) and the type of col is TIMESTAMP.
+
+If a partition column is defined by one of the preceding expressions, and a query filters data using the underlying base column of a generation expression, Delta Lake looks at the relationship between the base column and the generated column, and populates partition filters based on the generated partition column if possible. For example, given the following table:
+
+<CBCreateTable
+  python={
```
Comment (scottsand-db) [inline src/pages/learn/batch.mdx:388]: why only python right now? just curious. if we do want to add scala/sql, can you add a TODO in the code please.

**Review by scottsand-db — COMMENTED** [review]
Asked some questions. Code looks great to me, just some minor nits.

Also, as I asked in slack, how do I view + run this code myself? Can you include this in the PR description?

**Pull request closed without merging**